### PR TITLE
fix: add resources check on gateway-class in egctl translate

### LINKED ISSUE #1648: egctl x translate requires gateway/gatewayclass resources
*Description*:
`egctl x translate` panics when only given an httproute resource.

*Repro steps*:

<details>
<summary>Just httproute</summary>

```
» cat httproute.yaml      
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: app
spec:
  parentRefs:
    - name: default
      namespace: envoy-gateway-system
  hostnames:
    - test-auth.ihwa.liao.dev
  rules:
    - matches:
        - path:
            type: PathPrefix
            value: /
          headers:
            - name: authorization
              type: RegularExpression
              value: "Bearer .*"
      backendRefs:
        - name: web-app
          port: 80
    - matches:
        - path:
            type: PathPrefix
            value: /
      backendRefs:
        - name: auth-app
          port: 80

» ~/third_party/envoyproxy-gateway/default/egctl x translate --from gateway-api --to xds --type route --add-missing-resources --file httproute.yaml
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x1dabd3a]

goroutine 1 [running]:
github.com/envoyproxy/gateway/internal/cmd/egctl.addDefaultEnvoyProxy(0xc0007bdc80)
	github.com/envoyproxy/gateway/internal/cmd/egctl/translate.go:812 +0x1a
github.com/envoyproxy/gateway/internal/cmd/egctl.kubernetesYAMLToResources({0xc00083e500, 0x271}, 0x1)
	github.com/envoyproxy/gateway/internal/cmd/egctl/translate.go:801 +0x2334
github.com/envoyproxy/gateway/internal/cmd/egctl.translate({0x2750d40, 0xc0000b6030}, {0x7ffd4703fb68, 0xe}, {0x7ffd4703fb27, 0xb}, {0xc00037a270, 0x1, 0x1}, {0x24518dc, ...}, ...)
	github.com/envoyproxy/gateway/internal/cmd/egctl/translate.go:224 +0x159
github.com/envoyproxy/gateway/internal/cmd/egctl.NewTranslateCommand.func1(0xc0001ae400?, {0x2451af0?, 0x4?, 0x2451904?})
	github.com/envoyproxy/gateway/internal/cmd/egctl/translate.go:97 +0x14e
github.com/spf13/cobra.(*Command).execute(0xc000836300, {0xc0002158c0, 0x9, 0x9})
	github.com/spf13/cobra@v1.7.0/command.go:940 +0x87c
github.com/spf13/cobra.(*Command).ExecuteC(0xc0007c3800)
	github.com/spf13/cobra@v1.7.0/command.go:1068 +0x3a5
github.com/spf13/cobra.(*Command).Execute(0x0?)
	github.com/spf13/cobra@v1.7.0/command.go:992 +0x13
main.main()
	github.com/envoyproxy/gateway/cmd/egctl/main.go:16 +0x18
```

</details>

<details>
<summary>httproute + gatewayclass</summary>

```
» cat gatewayclass.yaml 
apiVersion: gateway.networking.k8s.io/v1beta1
kind: GatewayClass
metadata:
  name: envoyproxy
spec:
  controllerName: gateway.envoyproxy.io/gatewayclass-controller
---
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: app
spec:
  parentRefs:
    - name: default
      namespace: envoy-gateway-system
  hostnames:
    - test-auth.ihwa.liao.dev
  rules:
    - matches:
        - path:
            type: PathPrefix
            value: /
          headers:
            - name: authorization
              type: RegularExpression
              value: "Bearer .*"
      backendRefs:
        - name: web-app
          port: 80
    - matches:
        - path:
            type: PathPrefix
            value: /
      backendRefs:
        - name: auth-app
          port: 80

» ~/third_party/envoyproxy-gateway/default/egctl x translate --from gateway-api --to xds --type route --add-missing-resources --file gatewayclass.yaml 
{}
```

</details>

<details>
<summary>httproute + gateway + gatewayclass</summary>

```
» cat all.yaml         
apiVersion: gateway.networking.k8s.io/v1beta1
kind: GatewayClass
metadata:
  name: envoyproxy
spec:
  controllerName: gateway.envoyproxy.io/gatewayclass-controller
---
apiVersion: gateway.networking.k8s.io/v1beta1
kind: Gateway
metadata:
  name: default
  namespace: envoy-gateway-system
spec:
  gatewayClassName: envoyproxy
  listeners:
    - name: http
      protocol: HTTP
      port: 80
      allowedRoutes:
        namespaces:
          from: All
    - name: https
      protocol: HTTPS
      port: 443
      allowedRoutes:
        namespaces:
          from: All
      tls:
        mode: Terminate
        certificateRefs:
          - name: wildcard-letsencrypt
---
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: app
spec:
  parentRefs:
    - name: default
      namespace: envoy-gateway-system
  hostnames:
    - test-auth.ihwa.liao.dev
  rules:
    - matches:
        - path:
            type: PathPrefix
            value: /
          headers:
            - name: authorization
              type: RegularExpression
              value: "Bearer .*"
      backendRefs:
        - name: web-app
          port: 80
    - matches:
        - path:
            type: PathPrefix
            value: /
      backendRefs:
        - name: auth-app
          port: 80

» ~/third_party/envoyproxy-gateway/default/egctl x translate --from gateway-api --to xds --type route --add-missing-resources --file all.yaml
xds:
  envoy-gateway-system-default:
    '@type': type.googleapis.com/envoy.admin.v3.RoutesConfigDump
    dynamicRouteConfigs:
    - routeConfig:
        '@type': type.googleapis.com/envoy.config.route.v3.RouteConfiguration
        ignorePortInHostMatching: true
        name: envoy-gateway-system-default-http
        virtualHosts:
        - domains:
          - '*'
          name: envoy-gateway-system-default-http
          routes:
          - directResponse:
              status: 500
            match:
              headers:
              - name: :authority
                stringMatch:
                  exact: test-auth.ihwa.liao.dev
              - name: authorization
                stringMatch:
                  safeRegex:
                    regex: Bearer .*
              prefix: /
            name: envoy-gateway-system-app-rule-0-match-0-test-auth.ihwa.liao.dev
          - directResponse:
              status: 500
            match:
              headers:
              - name: :authority
                stringMatch:
                  exact: test-auth.ihwa.liao.dev
              prefix: /
            name: envoy-gateway-system-app-rule-1-match-0-test-auth.ihwa.liao.dev
```

</details>

*Environment*:
egctl built from current HEAD: e171678d87c13c89a48951878a2c5e7901f8de57

### PR DESCRIPTION
**Pull request by shawnh2**
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #1648

### CODE REVIEW

File: internal/cmd/egctl/translate.go [hunk 1]
```diff
@@ -357,8 +357,8 @@ func printOutput(w io.Writer, result TranslationResult, output string) error {
 	if err != nil {
 		return err
 	}
-	fmt.Fprintln(w, string(out))
-	return nil
+	_, ret := fmt.Fprintln(w, string(out))
```
Comment (Xunzhuo) [inline internal/cmd/egctl/translate.go:360]: why call it `ret` instead of reusing `err`?

**Review by Xunzhuo — COMMENTED** [review]
Can you add some UTs for this ?

File: internal/cmd/egctl/translate.go (hunk 1 above)
Comment (shawnh2) [reply internal/cmd/egctl/translate.go:360]: thought using `err` could cause a little confusing, but if not, we can still reuse `err`

**Comment by shawnh2** [comment]
> Can you add some UTs for this ?

sure thing

File: internal/cmd/egctl/translate.go (hunk 1 above)
Comment (shawnh2) [reply internal/cmd/egctl/translate.go]: fixed

**Comment by arkodg** [comment]
@shawnh2 I think there are 2 pieces associated with this issue
* Null Ptr Dereference here https://github.com/envoyproxy/gateway/blob/2edcafe430a2ac1e61ccb4d1a0f5a5ce00ad5a08/internal/cmd/egctl/translate.go#L812 which must be avoided with a check
* Should EG insert dummy `Gateway` and `GatewayClass` resources when `--add-default-resources` is set

**Comment by shawnh2** [comment]
> @shawnh2 I think there are 2 pieces associated with this issue
> 

1. we can add a check on nil pointer
2. in the [egctl doc](https://gateway.envoyproxy.io/latest/user/egctl.html#add-missing-resources), it specific that `--add-default-resources` is to use dummy **non Gateway API resources**, so if we decide to dummy Gateway and GatewayClass, we shall also change the doc. WDYT?

**Comment by arkodg** [comment]
@shawnh2 agreed on 2. 
I had added `EnvoyProxy` resource to it (which is a non Gateway API resource), but had forgotten to update the docs

**Comment by arkodg** [comment]
also can we pause on 2. and continue with 1. for now, would like @envoyproxy/gateway-maintainers to chime in on 2.

**Comment by shawnh2** [comment]
> also can we pause on 2. and continue with 1. for now, would like @envoyproxy/gateway-maintainers to chime in on 2.

sure, the change has been made, PTAL

File: internal/cmd/egctl/translate.go [hunk 2]
```diff
@@ -739,6 +739,11 @@ func kubernetesYAMLToResources(str string, addMissingResources bool) (*gatewayap
 		}
 	}
 
+	// Require the essential resources, like GatewayClass and Gateways
+	if resources.GatewayClass == nil || len(resources.Gateways) == 0 {
```
Comment (arkodg) [inline internal/cmd/egctl/translate.go:743]: is this needed ? now that you have added the check on L831

Comment (shawnh2) [reply internal/cmd/egctl/translate.go:743]: for now, if we use egctl x translate on only `httproute` + `gatewayclass`, we will get the output: `{}`.
is the output `{}` meet our expectations? if is, we can remove this check.

if not, we shoud return an error eg. `require gateway resource`, so we need to keep the `len(resources.Gateways) == 0` check, and as for the first one can be removed.

Comment (arkodg) [reply internal/cmd/egctl/translate.go:743]: yeah `{}` is fine for now, we want to mitigate the panic

and then we can decide on[ 2](https://github.com/envoyproxy/gateway/pull/1658#issuecomment-1641293073) based on community feedback

Comment (shawnh2) [reply internal/cmd/egctl/translate.go]: removed

also add some nil ptr check on gateway-class in `translateGatewayAPIToXds` and `translateGatewayAPIToGatewayAPI`, they all risk in panic

**Review by Xunzhuo — APPROVED** [review]
LGTM, thanks!

**Pull request merged**